src/legacy/ui/public/agg_types/param_types/base.ts

```diff
@@ -0,0 +1,90 @@
+/*
+ * Licensed to Elasticsearch B.V. under one or more contributor
+ * license agreements. See the NOTICE file distributed with
+ * this work for additional information regarding copyright
+ * ownership. Elasticsearch B.V. licenses this file to you under
+ * the Apache License, Version 2.0 (the "License"); you may
+ * not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *    http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+ * KIND, either express or implied.  See the License for the
+ * specific language governing permissions and limitations
+ * under the License.
+ */
+
+import { AggParam } from 'ui/agg_types';
+import { AggConfigs } from 'ui/vis/agg_configs';
+import { AggConfig } from '../../vis';
+import { SearchSource } from '../../courier';
+
+export class BaseParamType implements AggParam {
+  name: string;
+  type: string;
+  displayName: string;
+  required: boolean;
+  advanced: boolean;
+  editorComponent: any = null;
+  default: any;
+  write: (
+    aggConfig: AggConfig,
+    output: Record<string, any>,
+    aggConfigs?: AggConfigs,
+    locals?: Record<string, any>
+  ) => void;
+  serialize: (value: any) => any;
+  deserialize: (value: any, aggConfig: AggConfig) => any;
+  options: any[];
+  valueType?: any;
+
+  onChange?(agg: AggConfig): void;
+  shouldShow?(agg: AggConfig): boolean;
+  makeOrderAgg?(agg: AggConfig, state: unknown): AggConfig;
+
+  /**
+   *  A function that will be called before an aggConfig is serialized and sent to ES.
+   *  Allows aggConfig to retrieve values needed for serialization by creating a {SearchRequest}
+   *  Example usage: an aggregation needs to know the min/max of a field to determine an appropriate interval
+   *
+   *  @param {AggConfig} aggconfig
+   *  @param {Courier.SearchSource} searchSource
+   *  @param {Courier.SearchRequest} searchRequest
+   *  @returns {Promise<undefined>|undefined}
+   */
+  modifyAggConfigOnSearchRequestStart: (
+    aggconfig: AggConfig,
+    searchSource: SearchSource,
+    searchRequest: any
+  ) => void;
+
+  constructor(config: Record<string, any>) {
```

Comment:
The `config` taken by the constructor for each of the various ParamTypes is always `Record<string, any>` -- can't this be `AggParam` instead? It seems those are the only options that are provided, but I could be wrong.

Reply:
it could be for the base agg param type, others have additional options but i could have types for each of those. i think that could be done in the next step as this pr is already growing big.